File: binary_search/binarySearch.js


 function binarySearch(list, item){
   const size = list.length-1
   let [start, end] = [0, size]
   let res = null
   let step = 0 

   while(start<=end){

     step+=1
     let guess
     let middle = Math.floor((start+end)/2)

     guess = arr[middle]
     
     if(guess===item){
       res ={
         position:middle,
         step
       }
       break
     }
     if(guess>item){
       end = middle -1
     }else{
       start = middle+1
     }
     
                      
   }

   return res 
 }

const arr = [1,2,3,4,5,6,7,8,9]
console.log(binarySearch(arr,2)) // { position: 1, step: 2 }
console.log(binarySearch(arr,8)) // { position: 7, step: 3 } 
console.log(binarySearch(arr,25)) // null

File: binary_search/binary_search.py
def binary_search(my_list, item):
    start = 0
    end = len(my_list) - 1

    while start <= end:
        middle = (start + end) // 2
        guess = my_list[middle]

        if guess == item:
            return middle

        if guess > item:
            end = middle - 1
        else:
            start = middle + 1

    return None


my_list = [1, 2, 3, 4, 5, 6, 7, 8, 9]

print(binary_search(my_list, 2))  # 1
print(binary_search(my_list, 8))  # 7
print(binary_search(my_list, 25))  # none
